File: app/src/main/java/com/example/flowershop/database/dao/FlowerDao.java
package com.example.flowershop.database.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;
import com.example.flowershop.database.entity.Flower;

import java.util.List;

@Dao
public interface FlowerDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insert(Flower flower);

    @Update
    void update(Flower flower);

    @Delete
    void delete(Flower flower);

    @Query("SELECT * FROM Flower WHERE flowerID = :id")
    Flower getFlowerById(int id);

    @Query("SELECT * FROM Flower")
    LiveData<List<Flower>> getAllFlowers();

    @Query("SELECT * FROM Flower")
    List<Flower> getAllFlowersSync();

    @Query("SELECT * FROM Flower WHERE category = :category")
    LiveData<List<Flower>> getFlowersByCategory(String category);

    @Query("SELECT * FROM Flower WHERE category = :category")
    List<Flower> getFlowersByCategorySync(String category);

    @Query("SELECT * FROM Flower WHERE flowerName LIKE '%' || :name || '%'")
    LiveData<List<Flower>> searchFlowers(String name);
}

File: app/src/main/java/com/example/flowershop/database/dao/FlowerDao.kt
package com.example.flowershop.database.dao

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update
import com.example.flowershop.database.entity.Flower
import kotlinx.coroutines.flow.Flow

@Dao
interface FlowerDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(flower: Flower): Long

    @Update
    suspend fun update(flower: Flower)

    @Delete
    suspend fun delete(flower: Flower)

    @Query("SELECT * FROM Flower WHERE flowerID = :id")
    suspend fun getFlowerById(id: Int): Flower?

    @Query("SELECT * FROM Flower")
    fun getAllFlowers(): Flow<List<Flower>>

    @Query("SELECT * FROM Flower WHERE category = :category")
    fun getFlowersByCategory(category: String): Flow<List<Flower>>

    @Query("SELECT * FROM Flower WHERE flowerName LIKE '%' || :name || '%'")
    fun searchFlowers(name: String): Flow<List<Flower>>
}